Scripts/Runtime/Entities/Tasks/AbstractTaskDriverSystem.cs

```diff
@@ -12,89 +12,139 @@ namespace Anvil.Unity.DOTS.Entities
+            m_TaskDriverID++;
+            return id;
+        }
+
         protected JobTaskWorkConfig ConfigureUpdateJob(JobTaskWorkConfig.ScheduleJobDelegate scheduleJobDelegate)
         {
-            JobTaskWorkConfig config = new JobTaskWorkConfig(scheduleJobDelegate, this);
+            JobTaskWorkConfig config = new JobTaskWorkConfig(scheduleJobDelegate, this, SYSTEM_LEVEL_CONTEXT);
             m_UpdateJobData.Add(config);
             return config;
         }
-        
+
+        protected JobTaskWorkConfig ConfigureHandleCancelledJob(JobTaskWorkConfig.ScheduleJobDelegate scheduleJobDelegate)
+        {
+            JobTaskWorkConfig config = new JobTaskWorkConfig(scheduleJobDelegate, this, SYSTEM_LEVEL_CONTEXT);
+            m_HandleCancelledJobData.Add(config);
+            return config;
+        }
+
         //TODO: #39 - Some way to remove the update Job
 
-        protected VirtualData<TKey, TInstance> CreateData<TKey, TInstance>(params AbstractVirtualData[] sources)
-            where TKey : unmanaged, IEquatable<TKey>
-            where TInstance : unmanaged, IKeyedData<TKey>
+        protected VirtualData<TInstance> CreateData<TInstance>(VirtualDataIntent intent, params AbstractVirtualData[] sources)
+            where TInstance : unmanaged, IKeyedData
         {
-            VirtualData<TKey, TInstance> virtualData = VirtualData<TKey, TInstance>.Create(sources);
+            VirtualData<TInstance> virtualData = new VirtualData<TInstance>(intent, sources);
             m_InstanceDataLookup.AddData(virtualData);
+            
+            //TODO: Similarities on AbstractTaskDriver
+            //Add helper simple job to keep the data around each frame
+            if (intent == VirtualDataIntent.Persistent)
+            {
+                ConfigureUpdateJob(VirtualData<TInstance>.KeepPersistentJob.Schedule)
+                   .ScheduleOn(virtualData, BatchStrategy.MaximizeChunk)
+                   .RequireDataForUpdateAsync(virtualData);
+            }
             return virtualData;
         }
 
-        protected VirtualData<TKey, TInstance> GetData<TKey, TInstance>()
-            where TKey : unmanaged, IEquatable<TKey>
-            where TInstance : unmanaged, IKeyedData<TKey>
+        protected VirtualData<TInstance> GetData<TInstance>()
+            where TInstance : unmanaged, IKeyedData
         {
-            return m_InstanceDataLookup.GetData<TKey, TInstance>();
+            return m_InstanceDataLookup.GetData<TInstance>();
         }
 
         internal void RegisterTaskDriver(AbstractTaskDriver taskDriver)
         {
             Debug_EnsureTaskDriverSystemRelationship(taskDriver);
             m_TaskDrivers.Add(taskDriver);
         }
-        
+
         protected override void OnUpdate()
         {
+            JobTaskWorkConfig.Debug_EnsureNoDataLoss(this, m_InstanceDataLookup, m_UpdateJobData, ref m_HasCheckedUpdateJobsForDataLoss);
+            //TODO: DISCUSS - Should we ensure no CancelData loss?
```

Comment:
If you have Data on your TaskDriver or TaskDriverSystem that they own and it's not a `OneShot`, it needs to be updated and either resolved or written back to the Pending so it sticks around.

Otherwise you have data loss. 

We can trap against this with the Debug function which ensures we have a job scheduled to take care of it. If the data is Persistent we automatically create a `KeepPersistentJob` for it.

When data is cancelled though, we're currently not do anything about it from a safety perspective. 

You're welcome to schedule a HandleCancelJob so you can do something with the instances if you want but otherwise they just "poof" out of existence which is normally what you want. 

The question is whether we should enforce the developer to opt into a behaviour to explicitly deal with cancelled instances.

Reply:
Developer should be forced to explicitly define what the cancel behaviour is. Even if annoying it's better than forgetting to think through a cancel situation.

Back to my other question though, is there an actual optimization path to OneShot? If not, then let's just have every task required to write a result. Just like in the application code there's always a return code. I don't see the purpose to a task that you fire and forget but then for some reason want to be able to cancel.

Reply:
> Developer should be forced to explicitly define what the cancel behaviour is. Even if annoying it's better than forgetting to think through a cancel situation.

Ok sounds good.

> Back to my other question though, is there an actual optimization path to OneShot? If not, then let's just have every task required to write a result. Just like in the application code there's always a return code. I don't see the purpose to a task that you fire and forget but then for some reason want to be able to cancel.

There was an optimization but I think our other comments cover how the changes will negate or improve this.